libcache/cache.c

```diff
@@ -0,0 +1,654 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Cache library
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Malgorzata Wrobel
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cache.h"
+
+#include <sys/list.h>
+// #include "list.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+
+
+#define LIBCACHE_NUM_WAYS   4
+#define LIBCACHE_ADDR_WIDTH 64
+
+
+#define IS_VALID(f) (((f) & (1 << 0)) != 0 ? 1 : 0)
+#define IS_DIRTY(f) (((f) & (1 << 1)) != 0 ? 1 : 0)
+
+#define SET_VALID(f) \
+	do { \
+		(f) |= (1 << 0); \
+	} while (0)
+#define CLEAR_VALID(f) \
+	do { \
+		(f) &= ~(1 << 0); \
+	} while (0)
+
+#define SET_DIRTY(f) \
+	do { \
+		(f) |= (1 << 1); \
+	} while (0)
+#define CLEAR_DIRTY(f) \
+	do { \
+		(f) &= ~(1 << 1); \
+	} while (0)
+
+
+#define LOG2(x) ((uint8_t)(8 * sizeof(unsigned long) - __builtin_clzl((x)) - 1))
+
+
+typedef struct cacheline_s cacheline_t;
+
+
+struct cacheline_s {
+	uint64_t tag;
+	void *data;
+	uint64_t offset;
+	unsigned char flags;
+	cacheline_t *prev, *next; /* Circular doubly linked list */
+};
+
+
+typedef struct cacheset_s {
+	cacheline_t *timestamps;
+	cacheline_t *tags[LIBCACHE_NUM_WAYS];
+	cacheline_t lines[LIBCACHE_NUM_WAYS];
+	size_t count;
+} cacheset_t;
+
+
+struct cachectx_s {
+	cacheset_t *sets;
+
+	size_t lineSize;
+	size_t numLines;
+	size_t numSets;
+
+	uint64_t tagMask;
+	uint64_t setMask;
+	uint64_t offMask;
+
+	uint8_t offBitsNum;
+	uint8_t setBitsNum;
+	uint8_t tagBitsNum;
+
+	cache_readCb_t readCb;
+	cache_writeCb_t writeCb;
+};
+
+
+/* Generates mask of type uint64_t with numBits set to 1 */
+static uint64_t cache_genMask(int numBits)
+{
+	return ((uint64_t)1 << numBits) - 1;
+}
+
+
+cachectx_t *cache_init(size_t size, size_t lineSize, cache_writeCb_t writeCb, cache_readCb_t readCb)
+{
+	cachectx_t *cache = NULL;
+
+	if (size == 0 || lineSize == 0 || size < lineSize) {
```

Comment:
size < lineSize check is redundant - next if (size % lineSize != 0) will get that

Reply:
Thank you